### DISCUSSION TITLE
Feature Request: Add possibility to apply a variant for a class set with a delimiter

### DISCUSSION THREAD

**Discussion by SirMishaa** [tailwindlabs/tailwindcss#4679] [Ideas]
Hello! 

The variants are very useful, but sometimes a bit painful to write. This is especially the case when you want to apply several properties when hovering the element for example.

That's why I would like to propose a new syntax to specify the variant only once and to form a set.
From 
```html
<div>
  <button class="m-4 px-2 py-1 text-center bg-blue-500 text-white rounded-lg transform hover:scale-105 
  hover:bg-red-500 hover:font-semibold">A nice TailwindCSS button !</button>
</div>
```

To 
```html
<div>
  <button class="m-4 px-2 py-1 text-center bg-blue-500 text-white rounded-lg transform hover:{scale-105 
  bg-red-500 font-semibold}">A nice TailwindCSS button !</button>
</div>
```
The characters `{ }` form a group, the variant behind **will apply to all the elements of the group**.
This is easier to write and clearer.

We have more and more variants, I think it could be a little comfort, especially when there is a lot of property to change in a particular action.

**Comment by SirMishaa** [comment]
Blade is using ``{{ }}``, maybe another delimited should be used to avoid possible conflict between template engine.

**Reply by 3PSY0N** [reply]
I love this idea, but you are right, maybe use `hover:[ ]` or something like that, to avoid conflicts with some template renderers.

**Comment by SirMishaa** [comment]
@adamwathan 
Sorry for the inconvenience :p 
Is this something that you think is possible?

**Comment by VicGUTT** [comment]
Hey, yeah this idea has come up multiple times, usually with `()` as encapsulation.

I would imagine the Tailwindlabs team is not opposed to it but are more focused in ensuring the JIT engine is as steady as possible with current Tailwind's way of doings things before diving into more _exotic_ features.

We would also need to come up with a different _separator_ than spaces to differentiate between the encapsulated utilities, as spaces are not allowed in classes. So maybe something like :

```html
<button class="hover:{scale-105|bg-red-500|font-semibold}">
```

We shall see how TailwindCSS evolves over time :)

**Comment by moishinetzer** [comment]
Any updates on this idea, or if it's going to be on the roadmap?

**Comment by adamwathan** [comment]
Hey, can read about our explorations here:

https://twitter.com/adamwathan/status/1461519820411789314

We've got it working on a branch but not sure if we will include it or not because it's a performance anti-pattern.

**Reply by Bobakanoosh** [reply]
Why not make an opt-in feature, and make sure in the docs it notes that there can be a performance decrease because of it?

**Comment by rambhosale** [comment]
After going through the twitter thread one thing is clear 

This way of writing variant is easier, many want to have this feature but it affects the overall build size and negative affects the performance

How about implementing a emmet like completion in "tailwind autocomplete" which expands the the proposed new style into already existing structure instead of implementing this feature in tailwind core. That way it would be best of both worlds (to a extent). Easier to write and doesn't affect the build size. 

(I work mostly work on backend stuff forgive if there are any mistakes in terminologies)
